src/Enhancer/ConditionalEnhancer.php

```diff
@@ -48,12 +61,75 @@ public function __construct(array $enhancerMap)
      */
     public function enhance(array $defaults, Request $request)
     {
-        foreach ($this->enhancerMap as $pair) {
-            if ($pair['matcher']->matches($request)) {
-                return $pair['enhancer']->enhance($defaults, $request);
+        foreach ($this->getEnhancerMap() as $pair) {
+            if (($pair[self::MAP_KEY_MATCHER] instanceof RequestMatcherInterface
+                    && $pair[self::MAP_KEY_MATCHER]->matchRequest($request)
+                ) || $pair[self::MAP_KEY_MATCHER]->matches($request)) {
+                return $pair[self::MAP_KEY_ENHANCER]->enhance($defaults, $request);
             }
         }
 
         return $defaults;
     }
+
+    /**
+     * Creates a sorted list of enhancer map pairs. It is ordered by the position, which should be the key.
+     *
+     * @return array
+     */
+    public function getEnhancerMap()
+    {
+        if (count($this->enhancerMap) === 0 || count($this->sortedEnhancerMap) > 0) {
+            return $this->sortedEnhancerMap;
+        }
+
+        ksort($this->enhancerMap);
+        $this->sortedEnhancerMap = array();
+
+        foreach ($this->enhancerMap as $entries) {
+            foreach ($entries as $entry) {
+                $this->sortedEnhancerMap[] = $entry;
+            }
+        }
+
+        return $this->sortedEnhancerMap;
+    }
+
+    /**
+     * To create and add an enhancer map entry.
+     *
+     * @param string    $className FQCN of the enhancer.
+     * @param string    $source    Source value for the enhancer
+     * @param string    $target    Target value for the enhancer
+     * @param []|string $map       Map values for the enhancer
+     * @param []        $methods   List of methods the enhancer is restricted on.
+     * @param int       $position  To order the enhancer calls
+     */
+    public function createAndAddMapEntry($className, $source, $target, $map, $methods = null, $position = 0)
```

Comment:
imo this should just be addMapEntry and expect the caller to do the request matcher and stuff